Further tape and universal recorder improvements

cacogen opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

# About The Pull Request

_Ҝㄖﾌ丨爪卂 ᗪ乇ㄒ卂丨ㄥ_

### Flipping tapes
- Tapes have a second side that can be recorded to (20 minutes total)
- Tapes have a radial menu to choose between flipping or unspooling them (with shit icons)
- Tapes thrown have a 50% chance of flipping

### Capacity warning
- Tape recorder warns when tape is almost up
- Tape recorder shows amount of time left on tape in description (when not playing)

### SFX
#### Sources in .ogg file metadata (Freesound)
- Tape handling and flipping sound effects
- Tape recorder handling sound effects and play/record hiss loop

### Tape and transcript differentiation
- Tapes can be renamed to differentiate them from each other, like real world tapes
- Pens can no longer respool tapes (game ends up doing both at once)
- Tape recorder transcript names say they're of a tape or use the tape's name if it's not default
- Tapes have a random 3 character hexadecimal suffix to help differentiate them
- Random colour (i.e. randomly dyed greyscale) tapes added to random tape variant

### Chat message changes
- Tape recorder uses say() consistently instead of to_chat() or visible_message()
- Tape recorder has a speech span that's maroon on light and red on dark
- Tape recorder records itself saying logging messages instead of printing them to the transcript
- Tape recorder says "Tape empty." after playing an empty tape
- Tape recorder says "End of recording." before playback ends
- Tape recorder messages improved in misc. ways

### Misc. changes
- Tape recorder stops after playing or recording, instead of doing nothing sometimes (e.g. after pressing play on an empty tape)
- Tape recorder has 2 throwforce instead of 0 (should hurt slightly, which is how it was when it was added)
- Tape recorder no longer has unused radial menu icons (i.e. deletes them)
- Tape recorder Pause is now Stop (more accurate to its function and realism)
- Tape recorder won't print transcripts for empty tapes nor will the option show in the radial if you can't do it

TODO:
- [x] Speech spans
- [x] Change icon directory to one other radials use
- [x] Tape recorder drop and pick up sounds
- [x] Play or record hum loop
- [x] Tape flip icon mirror
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Makes it more real and more useful
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: cacogen
add: Tapes can now be turned over for double the recording time
tweak: Tapes can now be renamed by clicking them with a pen
del: Tapes can no longer be respooled using a pen (triggered renaming)
soundadd: Tape and tape recorder handling sounds
soundadd: Tape recorder play/record loop
tweak: Tape and tape recorder misc. improvements (see PR #56023 for details)
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

cacogen committed:
Further tape and universal recorder improvements

RaveRadbury commented:
Why not do a right click to rename instead of removing the respooling function?

cacogen committed:
Further further improvements

cacogen commented:
You can still use a screwdriver, to be clear. 

The UNIQUE_RENAME unintentional pen interactions (e.g. use pen to respool tape, then rename menu appears) is a separate issue I'll probably try to fix by adding pen clicking.

cacogen committed:
More stuff before I get a GitHub vacation

cacogen committed:
YEah

LemonInTheDark commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -22,6 +22,12 @@
 	var/list/icons_available = list()
 	var/icon_directory = 'icons/effects/icons.dmi'
+	///Whether we've warned during this recording that the tape is almost up.
+	var/seconds_warned = FALSE
```
This var name sounds more like the seconds since you warned someone, is that intended or just my brain playing tricks

cacogen replied:
Yeah it's hard to understand because of brevity I'll try to think of a better name

LemonInTheDark commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -30,10 +36,23 @@
 	update_icon()
 
+/obj/item/taperecorder/proc/readout()
+	var/readout = "<span class='notice'><b>NO TAPE INSERTED</b></span>"
+	if(mytape)
+		if(playing)
+			readout = "<span class='notice'><b>PLAYING</b></span>" //this should just tell you the time it's at but I can't figure it out
+		else
+			var/time = mytape.used_capacity
+			var/mins = round(time / 60)
+			var/secs = time - mins * 60
+			readout = "<span class='notice'><b>[mins]</b>m <b>[secs]</b>s</span>"
+	return readout
```
Early returns here? is there any point in  storing the text in readout if you're not going to do anything with it?

cacogen replied:
Well it returns readout, and the proc is embedded in the description of the tape recorder under examine().

LemonInTheDark replied:
Why not just return the text itself? seems like it could make things a little cleaner

cacogen replied:
Done

LemonInTheDark commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -138,12 +157,10 @@
 		playsound(src, 'sound/items/taperecorder/taperecorder_play.ogg', 50, FALSE)
-		recording = 1
+		recording = TRUE
+		say("<font color='[say_color]'>Recording started.</font>")
```
How drastic is the difference between this and \<span class='notice'\>?

cacogen replied:
Well it's maroon so pretty drastic. I think I could figure out how to make it use spans though.

LemonInTheDark commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -138,12 +157,10 @@
+		recording = TRUE
+		say("<font color='[say_color]'>Recording started.</font>")
 		update_icon()
-		mytape.timestamp += mytape.used_capacity
-		mytape.storedinfo += "\[[time2text(mytape.used_capacity * 10,"mm:ss")]\] Recording started."
```
Is this behavior gone?

cacogen replied:
It says in the PR body that this is replaced by it recording itself saying it rather than writing it directly.

cacogen committed:
Returns the text itself instead of a variable

cacogen committed:
Renames seconds_warned to time_left_warning for clarity

- Makes it so warning for time happens based on the time and not when the tape recorder hears something
- Makes stopping work correctly when we try to record with a full tape or when recording stops because of a full tape
- Uses the SECONDS and MINUTES defines instead of a mixture of undefined deciseconds and seconds that are multiplied by 10 (while having to divide by 10 in a couple of places)

cacogen committed:
Moves radial icons to proper directory

- Gives the icon file a better name

cacogen committed:
"End of recording." actually gets said at end of playback now

- Cannot print a transcript for a blank tape
- Tweaks transcript name setting to be better formatted

cacogen committed:
Transcript print no longer shows up in radial menu when you can't do it

cacogen committed:
"Tape empty." is said before stopping an empty tape now

- No longer update_icon() after stop() under play() (it's in the stop() proc)
- Reorders stopping to be consistent between playing and recording and logical (which probably makes no noticeable difference)

cacogen committed:
Print won't show in radial menu if you can't do it (attempt 2)

cacogen committed:
Tells you if the tape is empty when you try to play it (attempt 2)

cacogen committed:
Fixes fixing what it isn't broken

cacogen committed:
Gives tape recorder its own speech_span

cacogen committed:
Sounds

- Play/record hiss sound loop (mediocre due to lack of options in sound loops to cut and not fade in/out)
- Drop and pick up sounds
- 0→2 throwforce (like it was when Uhangi added them, before Hornygranny made it 0 for some reason)

cacogen committed:
Solves my icon flipping problem with flipped icons

- Cleans up radial image() usage
- Clarifies radial icon var
- Clarifies tape unspooling and respooling vars
- Random 3 char hexademical on tape names
- Tape can spawn on either side
- Adds greyscale tapes and randomised colouration

cacogen committed:
Dyed tape variant for use

cacogen commented:
Doing this sucked but I think I'm done now

LemonInTheDark commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -122,38 +144,44 @@
-			sleep(10)
-		recording = FALSE
-		update_icon()
+			mytape.used_capacity += 1 SECONDS
+			used += 1 SECONDS
+			if(max - used < time_left_warning && !time_warned)
+				time_warned = TRUE
+				say("[(max - used) / 10] seconds left!") //deciseconds / 10 = seconds
+			sleep(1 SECONDS)
+		if(used >= max)
+			say("Tape full.")
+		stop()
```
God I just noticed this, could you move this to a process? perhaps fast processing? then deal with stopping recording as an event rather then checking for it once a second?

cacogen replied:
I have no idea what that means and I didn't make the tape recorder so I'm not sure why you're mad at me about it.

LemonInTheDark replied:
Sorry if I came off that way, I'm not mad at you in the slightest.

This part of tape recorder code is operating off of a loop, which we then sleep to delay for some period of time. I dislike loops like this because it can fuck with our master-controller to some degree, and it means that we have less control of how each tick is spent in the game. I was asking you to move it to a subsystem, preferably SSfastprocess or maybe SSobj, along with moving this behavior and the readout below to a process() proc. As I said below, this is slightly out of scope so I can understand if you'd prefer not to.

cacogen replied:
Sorry, shouldn't have jumped to conclusions.

I'd love to make this less shit oldcode but I don't know how to do what you described and have little faith in my ability to figure it out by myself. Is there something you know of I could look at in the code that does what you described already?

LemonInTheDark replied:
Jared did something similar in this pr https://github.com/tgstation/tgstation/pull/55927. I recommend picking one hallucination and trying to parse out what's going on, preferably one of the shorter ones.

If you'd like more info on the concept/problem feel free to ask after me on discord, I appreciate the work you've done here already, and it'd be great if we could kill the jank dead.

LemonInTheDark commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -164,69 +192,66 @@
 			break
 		if(mytape.storedinfo.len < i)
+			say("End of recording.")
 			break
-		say(mytape.storedinfo[i])
+		say("[mytape.storedinfo[i]]")
 		if(mytape.storedinfo.len < i + 1)
 			playsleepseconds = 1
-			sleep(10)
-			say("End of recording.")
+			sleep(1 SECONDS)
 		else
 			playsleepseconds = mytape.timestamp[i + 1] - mytape.timestamp[i]
-		if(playsleepseconds > 14)
-			sleep(10)
-			say("Skipping [playsleepseconds] seconds of silence")
-			playsleepseconds = 1
+		if(playsleepseconds > 14 SECONDS)
+			sleep(1 SECONDS)
+			say("Skipping [playsleepseconds] seconds of silence.")
+			playsleepseconds = 1 SECONDS
 		i++
```
Maybe this too? I understand if it's a bit out of scope, but moving sleep loops to process is something we should do at some point

LemonInTheDark commented:
Note to self, rereview this

cacogen commented:
@LemonInTheDark 
I'm guessing this is probably going to be closed before you have a chance to. Mind removing the stale tag?

LemonInTheDark commented:
oh of course, sorry about that

cacogen commented:
Thanks

tgstation-server commented:
Your Fix/Feature pull request delta is currently below zero (-5). Maintainers may close future Feature/Tweak/Balance PRs. Fixing issues or helping to improve the codebase will raise this score.
